app/views/contests/scoreboard.html.erb

```diff
@@ -29,20 +39,25 @@
             (<%= @weighting[problem.id] %>)
           </th>
         <% end %>
-        <th style="text-align: right"> Score </th>
-        <th> Time </th>
-        <th> Rank </th>
+        <th style="width:1%; text-align: right"> Score </th>
+        <th style="width:1%"> Time </th>
       </tr>
     </thead>
     <tbody>
-      <% median = @scoreboard[@scoreboard.length/2-1] %>
       <% rank = 1 %>
+      <% num_unofficial = 0 %>
       <% previous_record = @scoreboard.first %>
       <% @scoreboard.each_with_index do |record,index| %>
-        <% if !user_signed_in? or record.user && record.user.id != current_user.id && !policy(@contest).inspect? %>
-          <% next if record.score < median.score || (record.score == median.score && record.time_taken > median.time_taken) # no permission to view %>
-        <% end %>
-        <tr <% if user_signed_in? && record.user && record.user.id == current_user.id %> class="emphasized"<% end %>>
+        <% next if !show_scoreboard && (!user_signed_in? or record.user && record.user.id != current_user.id) # no permission to view %>
+        <% is_unofficial = @contest.show_unofficial_contestants && (record.school.nil? || record.school_year.nil?) %>
+        <tr class="<% if user_signed_in? && record.user && record.user.id == current_user.id %>emphasized<% end %><% if is_unofficial %> unofficial<% end %>">
+          <% num_unofficial += 1 if is_unofficial %>
+          <% rank = index + 1 - num_unofficial unless previous_record.score == record.score && previous_record.time_taken == record.time_taken %>
+          <% previous_record = record %>
+          <td style="text-align: right"> <%= rank unless is_unofficial || !show_scoreboard %> </td>
```

Comment:
With the previous suggestion, this would become `rank if is_ranked && show_scoreboard`.